### PR TITLE
deploy: tweak enqueueing in the trigger controller

### PR DESCRIPTION
**Pull request by kargakis**
@mfojtik @smarterclayton

### CODE REVIEW

File: pkg/deploy/controller/generictrigger/factory.go [hunk 1]
```diff
@@ -89,24 +92,60 @@ func (c *DeploymentTriggerController) waitForSyncedStore(ready chan<- struct{},
...
+	if !deployutil.HasSynced(newDc, newDc.Generation) {
+		return
+	}
+	// Compare deployment config templates before enqueueing. This reduces the amount of times
+	// we will try to instantiate a deployment config to the exact number of times we need at
+	// the expense of duplicating some of the work that the instantiate endpoint is already doing
+	// but I think this is fine.
+	shouldInstantiate := true
+	if newDc.Status.LatestVersion > 0 {
+		latestRc, err := c.rcLister.ReplicationControllers(newDc.Namespace).Get(deployutil.LatestDeploymentNameForConfig(newDc))
```
Comment (smarterclayton) [inline pkg/deploy/controller/generictrigger/factory.go]: If the rc lister significantly lags, what happens here?

**Comment by smarterclayton** [comment]
[test]

**Comment by kargakis** [comment]
Updated to ignore errors from the rc cache and defer to instantiate to do the rc check. I opted for letting the dc controller update observedGeneration even when it observes that it cannot create a new rc so that the trigger controller will never be blocked.

File: pkg/deploy/controller/generictrigger/controller.go [hunk 2]
```diff
@@ -20,10 +21,14 @@ type DeploymentTriggerController struct {
...
-	// dcStoreSynced makes sure the dc store is synced before reconcling any deployment config.
-	dcStoreSynced func() bool
+	// dcLister provides a local cache for deployment configs.
+	dcLister oscache.StoreToDeploymentConfigLister
+	// dcListerSynced makes sure the dc store is synced before reconcling any deployment config.
+	dcListerSynced func() bool
+	// rcLister provides a local cache for replication controllers.
+	rcLister cache.StoreToReplicationControllerLister
+	// rcListerSynced makes sure the dc store is synced before reconcling any replication controller.
+	rcListerSynced func() bool
```
Comment (mfojtik) [inline pkg/deploy/controller/generictrigger/controller.go]: do you need this for testing?

File: pkg/deploy/controller/generictrigger/factory.go [hunk 3]
```diff
@@ -89,24 +92,64 @@ func (c *DeploymentTriggerController) waitForSyncedStore(ready chan<- struct{},
...
+		// If we get an error here it may be due to the rc cache lagging behind. In such a case
+		// just defer to the api server (instantiate REST) where we will retry this.
+		glog.V(2).Infof("Cannot get latest rc for dc %s:%d (%v) - will defer to instantiate", deployutil.LabelForDeploymentConfig(newDc), newDc.Status.LatestVersion, err)
+	} else {
+		initial, err := deployutil.DecodeDeploymentConfig(latestRc, c.codec)
+		if err != nil {
+			glog.V(2).Infof("Cannot decode dc from replication controller %s: %v", deployutil.LabelForDeployment(latestRc), err)
+			return
+		}
+		shouldInstantiate = !reflect.DeepEqual(newDc.Spec.Template, initial.Spec.Template)
```
Comment (mfojtik) [inline pkg/deploy/controller/generictrigger/factory.go]: how expensive is this?

File: pkg/deploy/controller/generictrigger/factory.go [hunk 4]
```diff
@@ -89,24 +92,64 @@ func (c *DeploymentTriggerController) waitForSyncedStore(ready chan<- struct{},
...
+	}
+	// Compare deployment config templates before enqueueing. This reduces the amount of times
+	// we will try to instantiate a deployment config at the expense of duplicating some of the
+	// work that the instantiate endpoint is already doing but I think this is fine.
+	shouldInstantiate := true
+	latestRc, err := c.rcLister.ReplicationControllers(newDc.Namespace).Get(deployutil.LatestDeploymentNameForConfig(newDc))
+	if err != nil {
+		// If we get an error here it may be due to the rc cache lagging behind. In such a case
+		// just defer to the api server (instantiate REST) where we will retry this.
+		glog.V(2).Infof("Cannot get latest rc for dc %s:%d (%v) - will defer to instantiate", deployutil.LabelForDeploymentConfig(newDc), newDc.Status.LatestVersion, err)
```
Comment (mfojtik) [inline pkg/deploy/controller/generictrigger/factory.go]: V(4)

File: pkg/deploy/controller/generictrigger/controller.go (hunk 2 above)
Comment (kargakis) [inline pkg/deploy/controller/generictrigger/controller.go]: nope, needed for the controller to know if its caches are synced

File: pkg/deploy/controller/generictrigger/factory.go [hunk 5]
```diff
@@ -89,24 +92,64 @@ func (c *DeploymentTriggerController) waitForSyncedStore(ready chan<- struct{},
...
+	} else {
+		initial, err := deployutil.DecodeDeploymentConfig(latestRc, c.codec)
+		if err != nil {
+			glog.V(2).Infof("Cannot decode dc from replication controller %s: %v", deployutil.LabelForDeployment(latestRc), err)
```
Comment (kargakis) [inline pkg/deploy/controller/generictrigger/factory.go]: We won't reconcile, I just return without adding the dc in the queue

File: pkg/deploy/controller/generictrigger/factory.go (hunk 3 above)
Comment (kargakis) [inline pkg/deploy/controller/generictrigger/factory.go]: As expensive as a deep equal check can be? We were running this in the trigger controller already prior to refactoring the controller to use the instantiate endpoint.

Comment (smarterclayton) [inline pkg/deploy/controller/generictrigger/factory.go]: Deep equal is expensive but it's the correct thing to use for now (vs something net new).

**Comment by mfojtik** [comment]
LGTM [merge]

**Comment by smarterclayton** [comment]
I guess it depends on what observedGeneration is supposed to mean.  Right
now it means "I've seen your request and I'm trying to bring it to
completion" but not "I've succeeded at reaching your request".

I don't think you can treat that the rc cache has anything to do with the
dc cache in this case though - the rc cache can wedge at any time and you
can be arbitrarily far behind the real state

**Comment by kargakis** [comment]
> I guess it depends on what observedGeneration is supposed to mean.  Right
> now it means "I've seen your request and I'm trying to bring it to
> completion" but not "I've succeeded at reaching your request".
> 
> I don't think you can treat that the rc cache has anything to do with the
> dc cache in this case though - the rc cache can wedge at any time and you
> can be arbitrarily far behind the real state

The rc cache has nothing to do with the dc cache, that's right. It will also be much heavier but my point with expecting a synced dc is that we have a better chance to not conflict with the dc controller.

**Pull request merged**